src/samplers/mh.jl

```diff
@@ -0,0 +1,249 @@
+  if is_first
+    push!(spl.info[:accept_his], true)
+    vi
+
+  else
+    if spl.alg.gid != 0 # Recompute joint in logp
+      runmodel(model, vi, nothing)
+    end
+    old_θ = copy(vi[spl])
+    old_logp = getlogp(vi)
+
+    dprintln(2, "Propose new parameters from proposals...")
+    propose(model, spl, vi)
+
+    dprintln(2, "computing accept rate α...")
+    α = getlogp(vi) - old_logp + spl.info[:proposal_ratio]
+
+    dprintln(2, "decide wether to accept...")
+    if log(rand()) < α && !spl.info[:violating_support]  # accepted
+      push!(spl.info[:accept_his], true)
+    else                      # rejected
+      push!(spl.info[:accept_his], false)
+      vi[spl] = old_θ         # reset Θ
+      setlogp!(vi, old_logp)  # reset logp
+    end
+
+    vi
+  end
+end
+
+function sample(model::Function, alg::MH;
+                chunk_size=CHUNKSIZE,     # set temporary chunk size
+                save_state=false,         # flag for state saving
+                resume_from=nothing,      # chain to continue
+                reuse_spl_n=0,            # flag for spl re-using
+                )
+
+  default_chunk_size = CHUNKSIZE  # record global chunk size
+  setchunksize(chunk_size)        # set temp chunk size
+
+  spl = reuse_spl_n > 0 ?
+        resume_from.info[:spl] :
+        Sampler(alg)
+  alg_str = "MH"
+
+  # Initialization
+  time_total = zero(Float64)
+  n = reuse_spl_n > 0 ?
+      reuse_spl_n :
+      alg.n_iters
+  samples = Array{Sample}(n)
+  weight = 1 / n
+  for i = 1:n
+    samples[i] = Sample(weight, Dict{Symbol, Any}())
+  end
+
+  vi = resume_from == nothing ?
+       model() :
+       deepcopy(resume_from.info[:vi])
+
+  if spl.alg.gid == 0
+    runmodel(model, vi, spl)
+  end
+
+  # MH steps
+  if PROGRESS spl.info[:progress] = ProgressMeter.Progress(n, 1, "[$alg_str] Sampling...", 0) end
+  for i = 1:n
+    dprintln(2, "$alg_str stepping...")
+
+    time_elapsed = @elapsed vi = step(model, spl, vi, i == 1)
+    time_total += time_elapsed
+
+    if spl.info[:accept_his][end]     # accepted => store the new predcits
+      samples[i].value = Sample(vi, spl).value
+    else                              # rejected => store the previous predcits
+      samples[i] = samples[i - 1]
+    end
+    samples[i].value[:elapsed] = time_elapsed
+
+    if PROGRESS ProgressMeter.next!(spl.info[:progress]) end
+  end
+
+  println("[$alg_str] Finished with")
+  println("  Running time        = $time_total;")
+  accept_rate = sum(spl.info[:accept_his]) / n  # calculate the accept rate
+  println("  Accept rate         = $accept_rate;")
+
+  setchunksize(default_chunk_size)      # revert global chunk size
+
+  if resume_from != nothing   # concat samples
+    unshift!(samples, resume_from.value2...)
+  end
+  c = Chain(0, samples)       # wrap the result by Chain
+  if save_state               # save state
+    save!(c, spl, model, vi)
+  end
+
+  c
+end
+
+function rand_truncated(dist, lowerbound, upperbound)
```

Comment:
Does this work for multivariable Gaussian as well?

Reply:
My rejection-free gaussian proposal implementation works only for univariate normal distributions (`Distributions.Normal{Float64}`). I used the derivations of this [blog post](http://fsaad.scripts.mit.edu/randomseed/metropolis-hastings-sampling-with-gaussian-drift-proposal-on-bounded-support/). I'm pretty sure that similar formulas can be derived for the multivariate case.

Reply:
It might be worth to extend this to multivariate Gaussian, or at least throw an error message saying not supported yet when `dist` is a multivariate Gaussian.

Reply:
I'll look how to extend this for the multivariate case, but not sure whether this is obvious or not.

Yet, that does not mean that proposing from a multivariate gaussian is not supported, but that if a proposed value (from the MVN) does not lie on the rv's support, the MH step will be rejected, i.e. this is not "proposal rejection-free".

Reply:
perhaps open a new issue for the multivariate Gaussian case and solve that in a separate PR?

Reply:
good idea

Reply:
good idea: https://github.com/yebai/Turing.jl/issues/387

Reply:
Is there any reason we have to write this customized function instead of using Distributions.jl's truncated Normal?